securedrop/journalist_app/api2/__init__.py

```diff
@@ -0,0 +1,158 @@
+        .options(joinedload(Source.submissions))
+        .options(joinedload(Source.replies))
+    )
+
+
+Version = NewType("Version", str)
+
+
+def json_version(d: dict) -> Version:
+    """
+    Calculate the version (BLAKE2s digest) of the normalized JSON representation
+    of the dictionary ``d``.
+
+    We use BLAKE2s here because SHA-256 is too slow (we don't care about
+    cryptographic security) and CRC-32 is too collision-prone (we're not merely
+    checksumming for transmission integrity).
+    """
+    s = json.dumps(d, separators=[",", ":"], sort_keys=True)
+    b = s.encode("utf-8")
+    return hashlib.blake2s(b).hexdigest()
+
+
+SourceUUID = NewType("SourceUUID", str)
+ItemUUID = NewType("ItemUUID", str)
+
+
+class IndexSourceEntry(TypedDict):
+    version: Version
+    collection: Dict[ItemUUID, Version]
+
+
+class Index(TypedDict):
+    sources: Dict[SourceUUID, IndexSourceEntry]
+
+
+@blp.get("/index")
+@blp.get("/index/<string:prefix>")
+def index(prefix: Optional[str] = None) -> Response:
+    """
+    By default, return the ETag-versioned ``Index`` consisting of an
+    ``IndexSourceEntry`` for each source and its collection, unless the client
+    provides the ETag of the current index.
+
+    Given a ``prefix``, return the sub-index of all sources whose UUIDs begin
+    with that prefix, unless the client provides the ETag of the current
+    sub-index for that prefix.  The client MAY choose an arbitrary prefix with
+    each request: e.g., a series of requests with the prefixes ``{0...f}`` will
+    effectively shard the index into 16 shards.
+    """
+    sources = {}
+
+    query = all_sources()
+    if prefix is not None:
+        if len(prefix) >= PREFIX_MAX_LEN:
+            abort(
+                422, f"malformed request; prefix must be shorter than {PREFIX_MAX_LEN} characters"
+            )
+
+        query = query.filter(Source.uuid.startswith(prefix))
+
+    for source in query.all():
+        all_source_metadata = source.to_api_v2()
+        source_entry: IndexSourceEntry = {
+            "version": json_version(all_source_metadata["source"]),
+            "collection": {},
+        }
+        for uuid, item in all_source_metadata["collection"].items():
+            source_entry["collection"][uuid] = json_version(item)
+        sources[source.uuid] = source_entry
+
+    index: Index = {"sources": sources}
+    version = json_version(index)
+    response = jsonify(index)
+
+    # If the request's `If-None-Match` header matches the version,
+    # return HTTP 304 with an empty response.
+    response.set_etag(version)
+    return response.make_conditional(request)
+
+
+class SourceDelta(TypedDict, total=False):
+    full_sources: List[SourceUUID]
+    partial_sources: Mapping[SourceUUID, List[ItemUUID]]
+
+
+class SourceEntry(TypedDict, total=False):
+    collection: Dict[ItemUUID, Any]
+    info: Optional[Mapping[str, Any]]  # omitted for partial sources
+
+
+class SourceMetadata(TypedDict):
```

Comment:
nit: maybe we could call this `SourcesResponse` or something of that nature? to me, `SourceMetadata` is the metadata for one source, so what would be in `SourceEntry.info`